src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -72,4 +88,240 @@ object RuleUtils {
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Replacement plan.
+   */
+  def getReplacementPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    if (ConfigUtils.getHybridScanEnabled(spark)) {
+      getHybridScanIndexPlan(spark, index, plan, useBucketSpec)
+    } else {
+      getIndexPlan(spark, index, plan, useBucketSpec)
+    }
+  }
+
+  /**
+   * Get alternative logical plan of the current plan using the given index.
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Alternative logical plan.
+   */
+  private def getIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    // Here we can't replace the plan completely with the index. This will create problems.
+    // For e.g. if Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
+    // if we replace this plan with index Scan (A,B,C), we lose the Filter(C=10) and it will
+    // lead to wrong results. So we only replace the base relation.
+    plan transformDown {
+      case baseRelation @ LogicalRelation(_: HadoopFsRelation, baseOutput, _, _) =>
+        val location =
+          new InMemoryFileIndex(spark, Seq(new Path(index.content.root)), Map(), None)
+        val relation = HadoopFsRelation(
+          location,
+          new StructType(),
+          StructType(index.schema.filter(baseRelation.schema.contains(_))),
+          if (useBucketSpec) Some(index.bucketSpec) else None,
+          new ParquetFileFormat,
+          Map())(spark)
+
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+  }
+
+  /**
+   * Get alternative logical plan of the current plan using the given index.
+   * With HybridScan, indexes with newly appended files to its source relation are also
+   * eligible and we reconstruct new plans for the appended files so as to merge into
+   * the index data.
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Alternative logical plan.
+   */
+  private def getHybridScanIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    val useBucketUnion = useBucketSpec || !index.relations.head.fileFormat.equals("parquet")
```

Comment:
Can you explain what the file format has to do it?

Reply:
We cannot read source files in a different format along with index data files (parquet). So need to find another way to read & merge the appended data.

In the case of FilterIndex, should we use `BucketUnion`? or is it okay to use just general `Union` as we don't use BucketSpec for FilterIndex now?
For instance, original plan: 
```
Project => Filter => Relation
```
BucketUnion:
```
BucketUnion => Project => Filter => Relation (index)
            |=> Shuffle => Project => Filter => Relation (appended data)
```
Union:
```
Union => Project => Filter => Relation (index)
      |=> Project => Filter => Relation (appended data)
```

Reply:
I see. I am also leaning toward using `Union` to eliminate shuffle. (We can measure this when you get a chance later). Btw, it would helpful if you add some comments on this.

Reply:
Ok then, can we go with (BucketUnion + BucketSpec) for the other formats for now and try `Union` with another PR?
I think it's okay because hybridScan config is false by default.

Reply:
Yes, sounds good to me. (just make sure to create an issue to follow up).

Reply:
Created #145 :)